### ISSUE TITLE
Rule proposal: Forbid tests with both a callback and a return statement

### ISSUE THREAD

**Issue by jfmengels** [lo1tuma/eslint-plugin-mocha#88]
In Mocha v3, it is not allowed to return a Promise in tests while the test function takes a `done` callback function.

It would be nice to detect this, so that you have a tighter feedback loop when coding, instead of waiting for your test to run. It could also help with the migration to v3 (I did this yesterday in a relatively big project, and it was a bit time-consuming).

I'm not sure whether we should forbid (top-level) `return` statements altogether, or only when we can detect that a Promise is returned (with a `.then` call). I think the former would be fine, as there is little value in returning something in a test when you have a callback, and it would cover a lot more cases where we can't know for sure that a Promise is being returned.

### Incorrect

```js
it('should XYZ', function(done) {
  return foo()
    .then(bar)
    .nodeify(done);
});
```

### Correct

```js
it('should XYZ', function(done) {
  foo()
    .then(bar)
    .nodeify(done);
});

it('should XYZ', function() {
  return foo()
    .then(bar);
});
```

As for the rule name, I'm not too sure yet. `no-return-and-callback`? Or if we want to make sure the returned statement is a Promise, `no-promise-and-callback`?

**Comment by lo1tuma** [comment]
I general I like this proposal. I think the hard part of this would be detect if a promise is returned or not. Do you know how mocha does this check? Does it fail for any non falsy value that is returned or only for objects that have a `then` method?

Maybe the easiest approach would be to warn on any top-level return (maybe with exceptions for `return;` `return undefined; and `return null;`). I think looking for the usage of calling a `then` method within the return statement is probably pretty error prone. Consider the following examples:

```js
it('foo', function () {
    // chai-as-promised assertions returns promises
    return expect(promiseReturningFunction()).to.be.fulfilled;
});
```

```js
it('foo', function () {
    return Promise.all([ checkFoo(), checkBar() ]);
});
```

**Comment by jfmengels** [comment]
> Does it fail for any non falsy value that is returned or only for objects that have a then method?

Only for Promises (so the latter). This is correct behavior for instance:
```js
describe('foo', function() {
  it('should', function(done) {
    function foo() {
      done();
    }
    setTimeout(foo, 500);
    return 3;
  });
});
```
But returning anything other than a Promise doesn't provide any value, so I'd say it's fine to warn if there is a return statement and a callback.

**Comment by lo1tuma** [comment]
> Not sure I see the point of having such statements in your tests anyway

Maybe it is used to skip further execution of the test, but then we would also need to support the following:

```js
it('foo', function (done) {
    foo(function (error, data) {
        if (error) {
            return done(error); // <-- return with the result of `done` call which is probably undefined
        }
        assert(true);
        done();
    });
});
```

**Comment by jfmengels** [comment]
> Maybe it is used to skip further execution of the test, but maybe you are right it is uncommon to do that anyway :wink: .

Ah, you mean something like this

```js
it('should xyz', function() {
  const res = foo();
  assert(res);
  if (!process.env.FOO) {
    return;
  }
  assert(res.xyz === 'foo');
});
```

Would make sense if you're testing stuff conditionally in an environment, or creating tests in a loop. You're right, I'll make the change later to ignore this when you return a nothing (`return;`) or a literal value (`return null/"foo"`) or undefined.

**Comment by jfmengels** [comment]
> You're right, I'll make the change later to ignore this when you return a nothing (return;) or a literal value (return null/"foo") or undefined.

That has been done in PR #94.
